// @redwoodjs/api/dist/auth/verifiers/common.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js/object/define-property");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault").default;

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.WebhookSignError = exports.WebhookVerificationError = exports.VERIFICATION_SIGN_MESSAGE = exports.VERIFICATION_ERROR_MESSAGE = exports.DEFAULT_WEBHOOK_SECRET = exports.verifierLookup = void 0;

var _jwtVerifier = _interopRequireDefault(require("./jwtVerifier"));

var _secretKeyVerifier = _interopRequireDefault(require("./secretKeyVerifier"));

var _sha1Verifier = _interopRequireDefault(require("./sha1Verifier"));

var _sha256Verifier = _interopRequireDefault(require("./sha256Verifier"));

var _skipVerifier = _interopRequireDefault(require("./skipVerifier"));

var _timestampSchemeVerifier = _interopRequireDefault(require("./timestampSchemeVerifier"));

var _process$env$WEBHOOK_;

const verifierLookup = {
  skipVerifier: _skipVerifier.default,
  secretKeyVerifier: _secretKeyVerifier.default,
  sha1Verifier: _sha1Verifier.default,
  sha256Verifier: _sha256Verifier.default,
  timestampSchemeVerifier: _timestampSchemeVerifier.default,
  jwtVerifier: _jwtVerifier.default
};
exports.verifierLookup = verifierLookup;
const DEFAULT_WEBHOOK_SECRET = (_process$env$WEBHOOK_ = process.env['WEBHOOK_SECRET']) !== null && _process$env$WEBHOOK_ !== void 0 ? _process$env$WEBHOOK_ : '';
exports.DEFAULT_WEBHOOK_SECRET = DEFAULT_WEBHOOK_SECRET;
const VERIFICATION_ERROR_MESSAGE = "You don't have access to invoke this function.";
exports.VERIFICATION_ERROR_MESSAGE = VERIFICATION_ERROR_MESSAGE;
const VERIFICATION_SIGN_MESSAGE = 'Unable to sign payload';
/**
 * Class representing a WebhookError
 * @extends Error
 */

exports.VERIFICATION_SIGN_MESSAGE = VERIFICATION_SIGN_MESSAGE;

class WebhookError extends Error {
  /**
   * Create a WebhookError.
   * @param {string} message - The error message
   * */
  constructor(message) {
    super(message);
  }

}
/**
 * Class representing a WebhookVerificationError
 * @extends WebhookError
 */


class WebhookVerificationError extends WebhookError {
  /**
   * Create a WebhookVerificationError.
   * @param {string} message - The error message
   * */
  constructor(message) {
    super(message || VERIFICATION_ERROR_MESSAGE);
  }

}
/**
 * Class representing a WebhookSignError
 * @extends WebhookError
 */


exports.WebhookVerificationError = WebhookVerificationError;

class WebhookSignError extends WebhookError {
  /**
   * Create a WebhookSignError.
   * @param {string} message - The error message
   * */
  constructor(message) {
    super(message || VERIFICATION_SIGN_MESSAGE);
  }

}
/**
 * VerifyOptions
 *
 * Used when verifying a signature based on the verifier's requirements
 *
 * @param {string} signatureHeader - Optional Header that contains the signature to verify
 * will default to DEFAULT_WEBHOOK_SIGNATURE_HEADER
 * @param {number} timestamp - Optional timestamp in msec
 * @param {number} tolerance - Optional tolerance in msec
 * @param {string} issuer - Options JWT issuer for JWTVerifier
 */


exports.WebhookSignError = WebhookSignError;
