// @redwoodjs/internal/dist/ast.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js/object/define-property");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault").default;

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.hasDefaultExport = exports.getCellGqlQuery = exports.getGqlQueries = exports.getNamedExports = exports.parse = void 0;

var _find = _interopRequireDefault(require("@babel/runtime-corejs3/core-js/instance/find"));

var _core = require("@babel/core");

var _parser = require("@babel/parser");

var _traverse = _interopRequireDefault(require("@babel/traverse"));

const parse = code => (0, _parser.parse)(code, {
  sourceType: 'module',
  plugins: ['jsx', 'typescript']
});

exports.parse = parse;

/**
 * get all the named exports in a given piece of code.
 */
const getNamedExports = code => {
  const namedExports = [];
  const ast = parse(code);
  (0, _traverse.default)(ast, {
    ExportNamedDeclaration(path) {
      var _path$node;

      // Re-exports from other modules
      // Eg: export { a, b } from './module'
      const specifiers = (_path$node = path.node) === null || _path$node === void 0 ? void 0 : _path$node.specifiers;

      if (specifiers.length) {
        for (const s of specifiers) {
          const id = s.exported;
          namedExports.push({
            name: id.name,
            type: 're-export'
          });
        }

        return;
      }

      const declaration = path.node.declaration;

      if (!declaration) {
        return;
      }

      if (declaration.type === 'VariableDeclaration') {
        const id = declaration.declarations[0].id;
        namedExports.push({
          name: id.name,
          type: 'variable'
        });
      } else if (declaration.type === 'FunctionDeclaration') {
        var _declaration$id;

        namedExports.push({
          name: declaration === null || declaration === void 0 ? void 0 : (_declaration$id = declaration.id) === null || _declaration$id === void 0 ? void 0 : _declaration$id.name,
          type: 'function'
        });
      } else if (declaration.type === 'ClassDeclaration') {
        var _declaration$id2;

        namedExports.push({
          name: declaration === null || declaration === void 0 ? void 0 : (_declaration$id2 = declaration.id) === null || _declaration$id2 === void 0 ? void 0 : _declaration$id2.name,
          type: 'class'
        });
      }
    }

  });
  return namedExports;
};
/**
 * get all the gql queries from the supplied code
 */


exports.getNamedExports = getNamedExports;

const getGqlQueries = code => {
  const gqlQueries = [];
  const ast = parse(code);
  (0, _traverse.default)(ast, {
    TaggedTemplateExpression(path) {
      const gqlTag = path.node.tag;

      if (gqlTag.type === 'Identifier' && gqlTag.name === 'gql') {
        gqlQueries.push(path.node.quasi.quasis[0].value.raw);
      }
    }

  });
  return gqlQueries;
};

exports.getGqlQueries = getGqlQueries;

const getCellGqlQuery = code => {
  const ast = parse(code);
  let cellQuery = undefined;
  (0, _traverse.default)(ast, {
    ExportNamedDeclaration({
      node
    }) {
      if (node.exportKind === 'value' && _core.types.isVariableDeclaration(node.declaration)) {
        var _context;

        const exportedQueryNode = (0, _find.default)(_context = node.declaration.declarations).call(_context, d => {
          return _core.types.isIdentifier(d.id) && d.id.name === 'QUERY' && _core.types.isTaggedTemplateExpression(d.init);
        });

        if (exportedQueryNode) {
          const templateExpression = exportedQueryNode.init;
          cellQuery = templateExpression.quasi.quasis[0].value.raw;
        }
      }

      return;
    }

  });
  return cellQuery;
};

exports.getCellGqlQuery = getCellGqlQuery;

const hasDefaultExport = code => {
  let exported = false;
  const ast = parse(code);
  (0, _traverse.default)(ast, {
    ExportDefaultDeclaration() {
      exported = true;
      return;
    }

  });
  return exported;
};

exports.hasDefaultExport = hasDefaultExport;
